### PR TITLE
feat(trino): parse BEGIN...END routine bodies with DECLARE/SET [CLAUDE]

### PR DESCRIPTION
**Pull request by rusackas**
Continuing with the next "episode" of our series. Thanks for the support ant thorough PR reviews thus far!

**PR 3 of ~7**, continuing #7934 and #7981.

Trino inline UDFs can use a `BEGIN...END` compound body instead of a bare `RETURN <expr>`, with `DECLARE`/`SET` statements and nested `BEGIN` blocks:

```sql
WITH FUNCTION meaning_of_life()
  RETURNS integer
  BEGIN
    DECLARE a integer DEFAULT 6;
    DECLARE b integer DEFAULT 7;
    RETURN a * b;
  END
SELECT meaning_of_life()
```

`DECLARE`/`SET` parsing and generation turned out to already be fully generic: `exp.Declare`/`DeclareItem`/`Set` already support Trino's exact grammar as-is, including the multi-identifier-one-type form (`DECLARE first_name, last_name, middle_name varchar(25)`). Trino only needed its own `DECLARE` tokenizer keyword and a `STATEMENT_PARSERS` entry (mirroring BigQuery/Spark/T-SQL), plus `DECLARE_DEFAULT_ASSIGNMENT = "DEFAULT"` (also matching BigQuery), since the base default is `"="`.

The routine body itself needed more care. `_parse_block()` (used for e.g. `CREATE PROCEDURE ... BEGIN ... END`) only recognizes its closing `END` when nothing follows it in the token stream — that doesn't fit Trino's grammar, where the UDF's `END` is always followed by the enclosing query with no separating semicolon (`... END SELECT f(1)`). `_parse_routine_block` reuses the same underlying chunk-advancing primitives (`_chunks`/`_advance_chunk`) that back `_parse_block`'s own semicolon/chunk-continuation handling (the mechanism that lets a routine body span the semicolon-delimited chunks `_parse()` splits the whole script into), but closes on `END` regardless of what follows it, leaving the remaining tokens for the caller. `block_sql` wraps unconditionally in `BEGIN ...`, since every `exp.Block` in Trino's grammar is one, including nested ones — a first pass only prefixed `BEGIN` at the top level and silently dropped it for nested blocks.

`_parse_routine_statement`'s fallback (for statements that aren't `BEGIN` or `RETURN`) dispatches through the same `STATEMENT_PARSERS` entries `_parse_statement()` uses, rather than delegating to `_parse_statement()` itself — that also carries a generic expression/`SELECT` fallback which would otherwise silently parse whatever follows a malformed routine body as if it were the routine's own body, instead of raising a clear error at the right place.

### Next steps (remaining follow-up PRs, each gated on the previous)

1. ~~Core `WITH FUNCTION ... RETURNS ... RETURN ...`~~ (#7934)
2. ~~Routine characteristics~~ (#7981)
3. **This PR** — `BEGIN...END` bodies, `DECLARE`/`SET`
4. `IF`/`ELSEIF`/`ELSE` — reusing/extending `exp.IfBlock`
5. `CASE...WHEN...END CASE`
6. `WHILE...DO...END WHILE` — reusing/extending `exp.WhileBlock` with a `label` arg
7. `LOOP`/`REPEAT`/`ITERATE`/`LEAVE`

Related: apache/superset#26162.

Disclosure: implemented with Claude Code, reviewed, tested (`make unit`, `make style`) and understood by me. Human me! I'm reeeeealllll.

### CODE REVIEW

**Comment by rusackas** [comment]
Spun up a real Trino instance (Docker, `trinodb/trino`) to verify the `test_inline_udf_begin_end` cases before anyone else had to catch it in review:

- `meaning_of_life()`, `one()` (the SET/counter example), the multi-identifier `DECLARE`, and `doubled(x)` combined with a real CTE all execute and return the expected values.
- Nested `BEGIN` actually runs (returns 2, not just a parse round-trip).
- The `LANGUAGE SQL + NOT DETERMINISTIC + BEGIN` combination itself is valid and runs fine (confirmed separately with a `RANDOM()` body) — the specific test case only fails on real Trino because its trivial body gets flagged by Trino's own function analysis as deterministic despite the `NOT DETERMINISTIC` declaration, the same class of issue already called out for `SECURITY`/`WITH (...)` in #7981. Updated the comment to say so.

**Comment by rusackas** [comment]
Also, just to layer on, wanted to say thanks for letting me onboard on the repo here. People on Superset have been patching around on things downstream, so I think there's a lot of opportunities to fix things upstream... I'll try to be selective, and not drive you all crazy, but let me know if I become a bother ;) I'm on the Slack now if you wanna chat.

File: sqlglot/dialects/trino.py [hunk 1]
```diff
@@ -15,6 +15,7 @@ class Tokenizer(Presto.Tokenizer):
         KEYWORDS = {
             **Presto.Tokenizer.KEYWORDS,
             "REFRESH": TokenType.REFRESH,
+            "DECLARE": TokenType.DECLARE,
```
Comment (fivetran-kwoodbeck) [inline sqlglot/dialects/trino.py:18]: `SELECT declare FROM (VALUES 1, 2) AS t(declare)` is valid Trino (and parses fine on main), but it now raises a `ParseError`.

Adding `TokenType.DECLARE` to the tokenizer changes how the word tokenizes everywhere.  Since `DECLARE` is not a reserved word in Trino,`TokenType.DECLARE` needs to be in `ID_VAR_TOKENS` / `TABLE_ALIAS_TOKENS`.

File: sqlglot/generators/trino.py [hunk 2]
```diff
@@ -63,6 +64,12 @@ def functionspecification_sql(self, expression: exp.FunctionSpecification) -> st
         body = self.sql(expression, "expression")
         return f"FUNCTION {self.sql(expression, 'this')}{characteristics_sql}{with_sql} {body}"
 
+    def block_sql(self, expression: exp.Block) -> str:
+        # Every exp.Block in Trino's grammar is a BEGIN...END block (nested ones
+        # included), so this wraps unconditionally rather than relying on callers
+        # to prefix BEGIN themselves.
+        return f"BEGIN {super().block_sql(expression)}"
```
Comment (fivetran-kwoodbeck) [inline sqlglot/generators/trino.py]: This assumes every `exp.Block` is a Trino routine body, but the generator can receives ASTs from other dialects, where the surrounding context already accounts for the `BEGIN`. Any such Block now renders with a doubled `BEGIN`:

```
import sqlglot
sqlglot.parse_one("BEGIN SELECT 1; SELECT 2; END", dialect="bigquery").sql("trino")
```
Behavior on main:
```
BEGIN SELECT 1; SELECT 2; END
```
This branch:
```
BEGIN BEGIN SELECT 1; SELECT 2; END
```

File: sqlglot/dialects/trino.py (hunk 1 above)
Comment (rusackas) [reply sqlglot/dialects/trino.py:18]: Good catch, confirmed. Turns out this exact gap is pre-existing for BigQuery/T-SQL/Spark too, none of them add their DECLARE keyword to ID_VAR_TOKENS either, so `SELECT declare FROM t` is already broken the same way there. Since it's Trino specifically where DECLARE isn't reserved, added it to Trino's own ID_VAR_TOKENS and TABLE_ALIAS_TOKENS rather than touching the shared base sets (which could be wrong for a dialect where it genuinely is reserved). Both your repro and a plain `SELECT declare FROM t` are covered by a new test now.

File: sqlglot/generators/trino.py (hunk 2 above)
Comment (rusackas) [reply sqlglot/generators/trino.py]: Good catch, and worth explaining precisely since it's not obvious. 

The BEGIN in your example's main output isn't coming from block_sql at all, it's preserved raw text inside a Command node (BigQuery's transaction parser only partially consumes "BEGIN SELECT 1" and falls back). My override was doubling text that was never mine to add in the first place. 

Fixed at the root instead of guessing by `isinstance`. I added a begin arg to the shared `exp.Block`, mirroring the identical flag `exp.Create` already carries for this same purpose, handled generically in the base `block_sql`, and set explicitly only `by _parse_routine_block` (the one place that should actually synthesize a BEGIN). 

No other dialect's `Block` ever sets it, so their output is unaffected, and confirmed against BigQuery/Snowflake/T-SQL/Spark's own tests plus your exact repro as a new test case on this PR.

File: sqlglot/dialects/trino.py (hunk 1 above)
Comment (rusackas) [reply sqlglot/dialects/trino.py:18]: Also, this might indicate a wider fix since it affects BigQuery/T-SQL/Spark, I can keep digging on a separate/new PR (or just an Issue filing) if you think it's worth more investigation.

File: tests/dialects/test_trino.py [hunk 3]
```diff
@@ -247,3 +247,57 @@ def test_inline_udf(self):
...
+
+        # DECLARE isn't reserved in Trino, so it must still work as a plain identifier
+        self.validate_identity("SELECT declare FROM (VALUES (1), (2)) AS t(declare)")
+
+        # A Block from another dialect's own parsing (not built by
+        # _parse_routine_block) must not get an extra synthesized BEGIN
+        self.assertEqual(
+            parse_one("BEGIN SELECT 1; SELECT 2; END", dialect="bigquery").sql(dialect="trino"),
+            "BEGIN SELECT 1; SELECT 2; END",
+        )
```
Comment (georgesittas) [inline tests/dialects/test_trino.py]: This test is a bit noisy because the first statement produces a `Command`:

Image

Perhaps we could just generate an AST programmatically and generate Trino SQL out of it:

```python
exp.Block(
    expressions=[
        parse_one("SELECT 1"),
        parse_one("SELECT 2"),
        exp.EndStatement()
    ]
).sql("trino")
```

File: sqlglot/dialects/trino.py (hunk 1 above)
Comment (georgesittas) [reply sqlglot/dialects/trino.py:18]: The following is valid in BigQuery, Spark, Presto, Trino, Athena: `SELECT 1 DECLARE`. It's only invalid in T-SQL.

File: sqlglot/parsers/trino.py [hunk 4]
```diff
@@ -13,6 +13,16 @@ class TrinoParser(PrestoParser):
...
+    STATEMENT_PARSERS = {
+        **PrestoParser.STATEMENT_PARSERS,
+        TokenType.DECLARE: lambda self: self._parse_declare(),
+    }
+
+    # DECLARE isn't reserved in Trino, so unlike the tokenizer keyword itself,
+    # it must still parse as a plain identifier/alias everywhere else.
+    ID_VAR_TOKENS = {*PrestoParser.ID_VAR_TOKENS, TokenType.DECLARE}
+    TABLE_ALIAS_TOKENS = {*PrestoParser.TABLE_ALIAS_TOKENS, TokenType.DECLARE}
+
```
Comment (georgesittas) [inline sqlglot/parsers/trino.py]: Given https://github.com/tobymao/sqlglot/pull/8004#discussion_r3693972605, I think I'd simply update `ID_VAR_TOKENS` in base. Then, you don't need to touch `TABLE_ALIAS_TOKENS` because the former dict expands within it. Also, in this way BigQuery, Presto, Athena and Spark automatically recognize `declare` as an identifier.

Moving the statement parser to base should be safe too, because it will only be visible to dialects that map `"DECLARE"` to the corresponding token, i.e., BigQuery, Trino etc.

Note: T-SQL will simply become more permissive, parse-wise, but I think that should be fine. We can double-check with its grammar and try some edge cases if we want, double-check what tests we already have, related to its `DECLARE` syntax, etc..

File: sqlglot/parsers/trino.py [hunk 5]
```diff
@@ -116,9 +126,53 @@ def _parse_function_specification(self) -> exp.FunctionSpecification:
             )
         )
 
+    def _parse_routine_block(self) -> exp.Block:
+        # _parse_block() only recognizes its closing END when nothing follows it,
+        # which doesn't fit Trino's grammar (the UDF's END is always immediately
+        # followed by the enclosing query: `... END SELECT f(1)`). This reuses the
+        # same chunk-advancing primitives (_chunks/_advance_chunk) _parse_block()
+        # relies on to thread a body across semicolon-delimited chunks, but closes
+        # on END unconditionally, leaving what follows it for the caller.
```
Comment (georgesittas) [inline sqlglot/parsers/trino.py]:
```suggestion
        # The body spans semicolon-delimited chunks like _parse_block(), but unlike it,
        # closes on END even when tokens follow; Trino's enclosing query always does
        # and is left for the caller (`... END SELECT f(1)`).
```

File: sqlglot/parsers/trino.py [hunk 6]
```diff
@@ -116,9 +126,53 @@ def _parse_function_specification(self) -> exp.FunctionSpecification:
...
+
+            if self._match(TokenType.END):
+                statements.append(exp.EndStatement())
+                break
+
+            statement = self._parse_routine_statement()
+            if not statement:
+                break
+
+            statements.append(statement)
```
Comment (georgesittas) [inline sqlglot/parsers/trino.py]:
```suggestion
        while not self._match(TokenType.END):
            if not self._curr:
                if self._chunk_index >= len(self._chunks):
                    self.raise_error("Unexpected end of routine body")
                    break

                self._advance_chunk()
            elif not self._match(TokenType.SEMICOLON):
                statement = self._parse_routine_statement()
                if not statement:
                    break

                statements.append(statement)
        else:
            statements.append(exp.EndStatement())
```

File: sqlglot/parsers/trino.py [hunk 7]
```diff
@@ -116,9 +126,53 @@ def _parse_function_specification(self) -> exp.FunctionSpecification:
...
             return self.expression(exp.Return(this=self._parse_disjunction()))
 
+        # SET/DECLARE are dispatched through the same STATEMENT_PARSERS entries
+        # _parse_statement() uses, but going through _parse_statement() itself would
+        # also enable its generic expression/SELECT fallback, silently swallowing
+        # whatever follows this routine's END as if it were the routine's own body.
+        if self._match_set(self.STATEMENT_PARSERS):
+            return self.STATEMENT_PARSERS[self._prev.token_type](self)
+
         self.raise_error("Expected routine statement")
```
Comment (georgesittas) [inline sqlglot/parsers/trino.py]: I'd rather we get rid of this whole highlighted section in favor of having explicit logic for `DECLARE`, `SET`:

```python
    ...
    # No need for weird comment that complicates readers :)
    if self._match(TokenType.DECLARE):
        return self._parse_declare()

    if self._match(TokenType.SET):
        return self._parse_set()
    ...
```

None of the statements you plan to add in the followup PRs (`WHILE`, `REPEAT`, etc) are handled by `STATEMENT_PARSERS`, so I think it'll be better to have explicit branches for them here. To that point, keeping the existing dispatch logic just for `DECLARE`, `SET` would be overkill, imo.

File: sqlglot/generator.py [hunk 8]
```diff
@@ -6231,7 +6231,8 @@ def chr_sql(self, expression: exp.Chr, name: str = "CHR") -> str:
 
     def block_sql(self, expression: exp.Block) -> str:
         expressions = self.expressions(expression, sep="; ", flat=True)
-        return f"{expressions}" if expressions else ""
+        begin = "BEGIN " if expression.args.get("begin") else ""
+        return f"{begin}{expressions}" if expressions else ""
```
Comment (georgesittas) [inline sqlglot/generator.py:6235]: Should this happen here or as an override in Trino?

**Review by georgesittas — CHANGES_REQUESTED** [review]

File: sqlglot/parsers/trino.py (hunk 4 above)
Comment (rusackas) [reply sqlglot/parsers/trino.py]: Done, moved TokenType.DECLARE into base ID_VAR_TOKENS/STATEMENT_PARSERS instead of a Trino override. Re-ran BigQuery/Presto/Athena/Spark/T-SQL's own suites to confirm, T-SQL is more permissive now as expected but nothing in its existing DECLARE tests broke.

File: sqlglot/parsers/trino.py (hunk 6 above)
Comment (rusackas) [reply sqlglot/parsers/trino.py]: Applied, thanks for the concrete rewrite. The while/else reads a lot cleaner than my original break-riddled version.

File: sqlglot/parsers/trino.py (hunk 7 above)
Comment (rusackas) [reply sqlglot/parsers/trino.py]: Fair, done. Dropped the STATEMENT_PARSERS lookup for explicit DECLARE/SET branches, matches how the WHILE/REPEAT branches will look too.

File: tests/dialects/test_trino.py (hunk 3 above)
Comment (rusackas) [reply tests/dialects/test_trino.py]: Good call, done. Built the Block directly instead of parsing through BigQuery's fallback, no more warning noise, and it's a more precise test of block_sql itself anyway.

File: sqlglot/generator.py (hunk 8 above)
Comment (rusackas) [reply sqlglot/generator.py:6235]: Went with base since exp.Create already carries an identical begin arg for this exact same purpose (a literal BEGIN a caller matched but doesn't itself preserve), handled generically in create_sql. This just extends the same pattern to Block rather than inventing a second one. Only Trino sets it today, but happy to make it a Trino-only override instead if you'd rather keep it scoped until something else actually needs it, let me know which way you lean.

**Comment by rusackas** [comment]
Pushed a round addressing everything above (I believe!). Thanks again for the thorough review:

- `TokenType.DECLARE` moved into base `ID_VAR_TOKENS`/`STATEMENT_PARSERS` instead of a Trino-only override, since it's valid as a plain identifier in BigQuery/Presto/Trino/Athena/Spark too, only T-SQL genuinely reserves it. Re-ran all five of those dialects' own suites to confirm nothing broke (T-SQL is intentionally more permissive now).
- `_parse_routine_block`'s loop restructured into the suggested while/else, and `_parse_routine_statement`'s STATEMENT_PARSERS-lookup fallback replaced with explicit `DECLARE`/`SET` branches.
- Rebuilt the `block_sql` regression test to construct the `Block` directly instead of parsing through BigQuery's `Command` fallback, dropping the unrelated warning noise.
- Replied inline on the `block_sql` base-vs-Trino-override question with my reasoning (matches `exp.Create`'s existing `begin` flag pattern), happy to go either way.

Full suite (1262 tests) green throughout. Happy to keep adjusting if needed!

File: sqlglot/dialects/trino.py (hunk 1 above)
Comment (rusackas) [reply sqlglot/dialects/trino.py:18]: Confirmed, thanks. Moved TokenType.DECLARE into base ID_VAR_TOKENS/STATEMENT_PARSERS (see the reply on your parsers/trino.py comment below) instead of a Trino-only fix, so BigQuery/Presto/Athena/Spark all pick it up too. Re-ran each of their suites to confirm, nothing broke.

**Review by georgesittas — APPROVED** [review]
Thank you!
